crates/bevy_ecs/src/world/archetype_invariants.rs

```diff
@@ -0,0 +1,358 @@
+    /// This may include empty archetypes: archetypes that contain no entities.
+    #[inline]
+    pub fn add(&mut self, archetype_invariant: UntypedArchetypeInvariant) {
+        self.raw_list.push(archetype_invariant);
+    }
+
+    /// Asserts that the provided iterator of [`ComponentId`]s obeys all archetype invariants.
+    ///
+    /// `component_ids` is generally provided via the `components` field on [`Archetype`](crate::archetype::Archetype).
+    ///
+    /// # Panics
+    ///
+    /// Panics if any archetype invariant is violated.
+    pub fn test_archetype(&self, component_ids_of_archetype: impl Iterator<Item = ComponentId>) {
+        let component_ids_of_archetype: HashSet<ComponentId> = component_ids_of_archetype.collect();
+
+        for invariant in &self.raw_list {
+            if invariant.predicate.test(&component_ids_of_archetype)
+                && !invariant.consequence.test(&component_ids_of_archetype)
+            {
+                let mut failed_invariants = vec![];
+
+                for invariant in &self.raw_list {
+                    if invariant.predicate.test(&component_ids_of_archetype)
+                        && !invariant.consequence.test(&component_ids_of_archetype)
+                    {
+                        failed_invariants.push(invariant.clone());
+                    }
+                }
+
+                panic!(
+                    "Archetype invariant violated! The following invariants were violated for archetype {:?}:\n{:?}",
+                    component_ids_of_archetype,
+                    failed_invariants,
+                )
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        self as bevy_ecs, component::Component, world::archetype_invariants::ArchetypeInvariant,
+        world::World,
+    };
+
+    #[derive(Component)]
+    struct A;
+
+    #[derive(Component)]
+    struct B;
+
+    #[derive(Component)]
+    struct C;
+
+    #[test]
+    fn full_bundle_happy() {
+        let mut world = World::new();
+
+        world.add_archetype_invariant(ArchetypeInvariant::<(A, B, C)>::atomic_bundle());
+        world.spawn().insert_bundle((A, B, C));
+    }
+
+    #[test]
+    fn full_bundle_on_insert_happy() {
+        let mut world = World::new();
+
+        world.spawn().insert_bundle((A, B, C));
+        world.add_archetype_invariant(ArchetypeInvariant::<(A, B, C)>::atomic_bundle());
+    }
+
+    #[test]
+    #[should_panic]
+    fn full_bundle_sad() {
+        let mut world = World::new();
+
+        world.add_archetype_invariant(ArchetypeInvariant::<(A, B, C)>::atomic_bundle());
+        world.spawn().insert_bundle((A, B));
+    }
+
+    #[test]
+    #[should_panic]
+    fn full_bundle_on_insert_sad() {
+        let mut world = World::new();
+
+        world.spawn().insert_bundle((A, B));
+        world.add_archetype_invariant(ArchetypeInvariant::<(A, B, C)>::atomic_bundle());
+    }
+}
```

Comment:
Renaming `full_bundle` to `atomic_bundle` in tests.

```suggestion
    #[test]
    fn atomic_bundle_happy() {
        let mut world = World::new();

        world.add_archetype_invariant(ArchetypeInvariant::<(A, B, C)>::atomic_bundle());
        world.spawn().insert_bundle((A, B, C));
    }

    #[test]
    fn atomic_bundle_on_insert_happy() {
        let mut world = World::new();

        world.spawn().insert_bundle((A, B, C));
        world.add_archetype_invariant(ArchetypeInvariant::<(A, B, C)>::atomic_bundle());
    }

    #[test]
    #[should_panic]
    fn atomic_bundle_sad() {
        let mut world = World::new();

        world.add_archetype_invariant(ArchetypeInvariant::<(A, B, C)>::atomic_bundle());
        world.spawn().insert_bundle((A, B));
    }

    #[test]
    #[should_panic]
    fn atomic_bundle_on_insert_sad() {
        let mut world = World::new();

        world.spawn().insert_bundle((A, B));
        world.add_archetype_invariant(ArchetypeInvariant::<(A, B, C)>::atomic_bundle());
    }
}
```